### DISCUSSION TITLE
Fail when a mock is called outside of exhaustive verify

### DISCUSSION THREAD

**Discussion by Cedrick84** [lupuuss/Mokkery#23] [Q&A] [answered]
Hi all,

I am new to Mokkery and like the ease of use so far! 

To make my tests really strict I would like to achieve the following.
Consider the following code.

```
val firstMock = mock<FirstDependency>
val secondMock = mock<SecondDependency> 

val sut = SUT(firstMock, secondMock)
sut.methodThatCallsBothMocks()

verify(exhaustive) { 
  firstMock.method()
}
```

Even though I call a method that invokes first and second mock, the test passes. I expect it to fail as a second dependency is called but this method is not being verified. Is there any way to achieve this behaviour?

Many thanks!

**Comment by lupuuss** [comment]
Hi! `verify` only includes mocks that are invoked inside its block. To verify that some mocks were not called at all you can use [verifyNoMoreCalls](https://mokkery.dev/api_reference/mokkery-runtime/dev.mokkery/verify-no-more-calls.html).

**Reply by Cedrick84** [reply]
This does work, however now I need to explicitly invoke `verifyNoMoreCalls` on all my dependencies. In my production code I have a lot more dependencies than this. It is also easy to miss adding a  `verifyNoMoreCalls`  to each of the test cases of a specific SUT.
I am looking for a solution where anything missed in the verify block causes a failure, instead of explicitly verifying everything.

**Reply by lupuuss** [reply, marked as answer]
I know what you mean. However, implementing this check automatically is problematic. For a reference, Mockk acts the same in given scenario. Maybe with compiler plugin capabilities it could be possible, but it's not easy to do and probably error-prone. 
Also, you can eliminate partially the boilerplate with `AfterTest`:
```kotlin
@AfterTest
fun after() {
   verifyNoMoreCalls(dep1, dep2, dep3)
}
```

**Reply by Cedrick84** [reply]
Thank you so much @lupuuss , I will follow that advice for now!
